Remove all use of deprecated coordinate transform constructors

elpaso opened the pull request:
... and ensure all coordinate transforms correctly utilise the project's coordinate transform context.

Sponsored by ICSM

elpaso committed:
Add coordinate transform context to data provider options

First step to pass the context down into the rabbit hole.

elpaso committed:
Step2 of coordinate traform context passing: SIP broken

elpaso committed:
Fix SIP (sipify does not like inline initializers)

elpaso committed:
Fix deprecated PUSH/POP

elpaso committed:
Remove all deprecated usages of QgsCoordinateTransform without context

elpaso committed:
Remove unused fwd declaration

elpaso committed:
Handle the coordinate tranform context in the layer ctor

Also use the context when read/write XML by passing it
in the R/W context

elpaso committed:
coordinateTransformContext -> transformContext for consistency

I like "coordinateTransformContext" more btw.

elpaso committed:
More coordinateTranformContext -> tranformContext

elpaso committed:
Remove deprecated QgsCoordinateTransform ctor ...

... and add a getter for coordinateTransform to QgsMapLayer

The reason is that I think we should reduce the dataProvider API
usage and rely on the higher level QgsMapLayer API whenever it
is possible, QgsMapLayer checks for data provider validity
and returns a default constructed QgsCoordinateTransform instead
of crashing.

elpaso committed:
Dox typos

nyalldawson commented on src/core/mesh/qgsmeshlayer.h:
```diff
@@ -97,7 +97,15 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
+      /**
+       * Constructor for LayerOptions.
+       */
+      explicit LayerOptions( const QgsCoordinateTransformContext &transformContext = QgsCoordinateTransformContext() )
```
I'd remove the default QgsCoordinateTransformContext() here -- we want to force people to specify a correct one, and not give them any opportunity to just use a default constructed context.

nyalldawson commented on src/core/processing/qgsprocessingutils.cpp:
```diff
@@ -639,9 +639,7 @@ QgsRectangle QgsProcessingUtils::combineLayerExtents( const QList<QgsMapLayer *>
-      Q_NOWARN_DEPRECATED_PUSH
-      QgsCoordinateTransform ct( layer->crs(), crs );
-      Q_NOWARN_DEPRECATED_POP
+      QgsCoordinateTransform ct( layer->crs(), crs, QgsProject::instance()->transformContext() );
```
Avoid QgsProject::instance here -- instead, add a QgsProcessingContext argument to the function and use the QgsProcessingContext's transformContext instead.

nyalldawson commented on src/core/qgsvectorlayer.h:
```diff
@@ -401,9 +401,12 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
        * Constructor for LayerOptions.
        */
-      explicit LayerOptions( bool loadDefaultStyle = true, bool readExtentFromXml = false )
+      explicit LayerOptions( bool loadDefaultStyle = true,
```
Can you deprecate this constructor, and make a new one which requires an explicit QgsCoordinateTransformContext value ? (We don't want ANY default constructed contexts)

elpaso replied:
Yes, but what do we do with 
`explicit QgsVectorLayer( ... const QgsVectorLayer::LayerOptions &options = QgsVectorLayer::LayerOptions()` ?

nyalldawson commented on src/providers/arcgisrest/qgsafsprovider.cpp:
```diff
@@ -117,9 +117,7 @@ QgsAfsProvider::QgsAfsProvider( const QString &uri, const ProviderOptions &optio
-    Q_NOWARN_DEPRECATED_PUSH
-    mSharedData->mExtent = QgsCoordinateTransform( extentCrs, mSharedData->mSourceCRS ).transformBoundingBox( mSharedData->mExtent );
-    Q_NOWARN_DEPRECATED_POP
+    mSharedData->mExtent = QgsCoordinateTransform( extentCrs, mSharedData->mSourceCRS, transformContext() ).transformBoundingBox( mSharedData->mExtent );
```
use the transform context from ProviderOptions options here.

nyalldawson commented on src/providers/db2/qgsdb2provider.h:
```diff
@@ -101,7 +101,7 @@ class QgsDb2Provider : public QgsVectorDataProvider
       QMap<int, int> *oldToNewAttrIdxMap,
       QString *errorMessage = nullptr,
-      const QMap<QString, QVariant> *options = nullptr
+      const QMap<QString, QVariant> *coordinateTransformContext = nullptr
```
What's this for?

nyalldawson commented:
Looking good, and definitely on the right track.

What we need to avoid is storing the transform context wherever possible (i.e. just use transient values to update any stored QgsCoordinateTransforms, and then discard the context).

So we need to remove the getters from map layer/data provider. I realise this effects the vector layer writer/exporter and raster writer classes, so these will need to be handled by adding setters for the transform context to those 3 classes (and deprecating the api for those classes which DON'T have the context specified)... and then let's see what build warnings that exposes, and work out how to address those (I realise it's going to open up more code we need to fix).

elpaso committed:
Removed leftover

elpaso committed:
Use options from provider ctor

elpaso committed:
Deprecate vector layer ctor with default constructed options

replaced by QgsVectorLayer( const QgsVectorLayer::LayerOptions &options,
...

elpaso committed:
Add mandatory options to QgsVectorLayer ctor

... and deprecate the rest.

In QGIS 4 we'll have to get rid of the deprecated default ctor and
create a new one which takes no arguments.

elpaso commented:
@nyalldawson  100 changed files sound like a good number ;) 

Seriously, I've battled with `QgsVectorLayer` (default) ctor and before I apply the same solution to the raster and other layers I'd like you to have a look.

The issue here was that in order to default construct the `QgsVectorLayer` (which is apparently also required by QML) we used a default constructed `QgsVectorLayer::LayerOptions` that we don't have anymore, now an instance of `QgsVectorLayer::LayerOptions` is always required when constructing a vector layer (I will do the same with other layers if that's ok).

There are many places in the code where a `QgsVectorLayer` is constructed, and in order to get the `transformContext` I often didn't have any other choice than to get it from the `QgsProject::instance()`, I know it's not ideal and we are trying to get rid of it: please tell me if you think that I should change all the API calls that construct a `QgsVectorLayer` to accept a `QgsCoordinateTransformContext` instance, this will probably require the deprecation of many more API calls.

elpaso committed:
Deprecate method without transformContext

elpaso committed:
Deprecate default ctor for mesh layer options

nyalldawson committed:
Tweak QgsProcessingUtils::combineLayerExtents for future proofing, remove deprecated usage

nyalldawson committed:
Fix use of deprecated/removed coordinate transform constructor

nyalldawson committed:
Some build fixes

elpaso committed:
Fix test build

elpaso committed:
Restore default ctor for options and layer

elpaso committed:
doxy

elpaso committed:
Do not store context in layer (but in provider) ...

... also use transform context in raster projector
and deprecate setCrs without context

elpaso committed:
Raster transform context

elpaso committed:
Fix class documentation

elpaso committed:
Missing doc in an unrelated class :(

elpaso committed:
More doxy

elpaso committed:
Remove build warnings

elpaso committed:
Fix grass alg combineLayerExtents

elpaso committed:
Remove test case (API is not public anymore)

nyalldawson commented on src/analysis/raster/qgsrastercalculator.cpp:
```diff
@@ -176,7 +176,7 @@ QgsRasterCalculator::Result QgsRasterCalculator::processCalculation( QgsFeedback
         {
           QgsRasterProjector proj;
-          proj.setCrs( ref.raster->crs(), mOutputCrs );
+          proj.setCrs( ref.raster->crs(), mOutputCrs, ref.raster->dataProvider()->transformContext() );
```
Instead of taking this from the dataProvider(), I think it'd be cleaner to add it in the QgsRasterCalculator constructor (and *sigh* deprecate the other existing ones)

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -7504,7 +7505,7 @@ QString QgisApp::saveAsRasterFile( QgsRasterLayer *rasterLayer, const bool defau
     {
       QgsRasterProjector *projector = new QgsRasterProjector;
-      projector->setCrs( rasterLayer->crs(), d.outputCrs() );
+      projector->setCrs( rasterLayer->crs(), d.outputCrs(), rasterLayer->dataProvider()->transformContext() );
```
Here I'd take it from the project instance again -- just on the odd case that the data provider is out of sync (and seeing that we can safely use the project instance here)

nyalldawson commented on src/core/layout/qgslayoutitemmapoverview.cpp:
```diff
@@ -33,7 +33,7 @@
 QgsLayoutItemMapOverview::QgsLayoutItemMapOverview( const QString &name, QgsLayoutItemMap *map )
   : QgsLayoutItemMapItem( name, map )
-  , mExtentLayer( qgis::make_unique< QgsVectorLayer >( QStringLiteral( "Polygon?crs=EPSG:4326" ), QStringLiteral( "overview" ), QStringLiteral( "memory" ) ) )
+  , mExtentLayer( qgis::make_unique< QgsVectorLayer >( QStringLiteral( "Polygon?crs=EPSG:4326" ), QStringLiteral( "overview" ), QStringLiteral( "memory" ), QgsVectorLayer::LayerOptions( QgsCoordinateTransformContext() ) ) )
```
```suggestion
  , mExtentLayer( qgis::make_unique< QgsVectorLayer >( QStringLiteral( "Polygon?crs=EPSG:4326" ), QStringLiteral( "overview" ), QStringLiteral( "memory" ), QgsVectorLayer::LayerOptions( map && map->layout() && map->layout()->project() ? map->layout()->project()->transformContext() : QgsCoordinateTransformContext() ) ) )
```

nyalldawson commented on src/core/mesh/qgsmeshlayer.h:
```diff
@@ -218,6 +228,16 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
+     *
+     * \since QGIS 3.8
+     */
+    virtual void setTransformContext( const QgsCoordinateTransformContext &transformContext ) override;
```
```suggestion
    void setTransformContext( const QgsCoordinateTransformContext &transformContext ) override;
```

nyalldawson commented on src/core/qgsmaplayer.h:
```diff
@@ -612,6 +612,14 @@ class CORE_EXPORT QgsMapLayer : public QObject
+    /**
+     * Returns the layer data provider coordinate transform context
+     * or a default transform context if the layer does not have a valid data provider.
+    \since QGIS 3.8
```
```suggestion
    * \since QGIS 3.8
```

nyalldawson commented on src/core/qgsofflineediting.cpp:
```diff
@@ -235,8 +235,8 @@ void QgsOfflineEditing::synchronize()
     remoteName.remove( QRegExp( " \\(offline\\)$" ) );
-
-    QgsVectorLayer *remoteLayer = new QgsVectorLayer( remoteSource, remoteName, remoteProvider );
+    const QgsVectorLayer::LayerOptions options { QgsProject::instance()->transformContext() };
```
Given the amount of QgsProject::instance() used elsewhere here, this is fine ;)

nyalldawson commented on src/core/raster/qgsrasterfilewriter.cpp:
```diff
@@ -231,9 +231,7 @@ QgsRasterFileWriter::WriterError QgsRasterFileWriter::writeDataRaster( const Qgs
-        Q_NOWARN_DEPRECATED_PUSH
-        QgsCoordinateTransform ct( projector->destinationCrs(), projector->sourceCrs() );
-        Q_NOWARN_DEPRECATED_POP
+        QgsCoordinateTransform ct( projector->destinationCrs(), projector->sourceCrs(), srcProvider->transformContext() );
```
I think it'd be cleaner if this was set in QgsRasterFileWriter::writeRaster

elpaso replied:
@nyalldawson ok, this class qgsrasterfilewriter.cpp is like the Pandora's box. 
There are raster provider factories that do not accept raster options and do not set the transform context. 
What do you suggest to do here? Should it be responsibility of the callers to set the context in the provider?

nyalldawson replied:
I might be misunderstanding the issue -- but I don't think these factories will need the context. It's only the pipe with the optional transformer which would need them. (Otherwise can you point to the troublesome code?)

elpaso replied:
@nyalldawson the potential issue here is that, in general, the new approach is to construct a raster layer by passing LayerOptions that contains the tc and store it into the provider (because we need it for cloning), the factories are constructing a data provider that do not set the tc, so the caller will be responsible for setting it when the data provider is used to construct a layer. We can leave it as it is but I thought it was important to discuss it.

Additionally passing the tc in `writeRaster` is not enough, because it uses `writeImageRaster` that requires a context, but this is not an issue.

See if the last commit works.

nyalldawson replied:
Looks good to me, are you happy?

elpaso replied:
Happy like Easter ... (Italian way of saying that means very happy)

nyalldawson commented on src/core/raster/qgsrasterprojector.cpp:
```diff
@@ -747,14 +758,14 @@ QgsRasterBlock *QgsRasterProjector::block( int bandNo, QgsRectangle  const &exte
   if ( feedback && feedback->isCanceled() )
     return new QgsRasterBlock();
 
+  // TODO: check if the last condition always apply to context
```
Yep, it's stiill safe to shortcut here

nyalldawson commented on src/core/raster/qgsrasterprojector.h:
```diff
@@ -69,9 +69,19 @@ class CORE_EXPORT QgsRasterProjector : public QgsRasterInterface
+     * Sets the source and destination CRS
+     * \deprecated since QGIS 3.8, use transformContext version instead
+     */
+    Q_DECL_DEPRECATED void setCrs( const QgsCoordinateReferenceSystem &srcCRS, const QgsCoordinateReferenceSystem &destCRS,
```
Thanks!

nyalldawson commented on python/core/auto_generated/qgsreadwritecontext.sip.in:
```diff
@@ -111,6 +111,24 @@ Sets the project translator. Means it shouldn't conform mDefaultTranslator anymo
 .. versionadded:: 3.4
+%End
+
+    QgsCoordinateTransformContext transformContext() const;
```
Are these additions still necessary? I gather they aren't currently used here

elpaso replied:
`transformContext` it's used in `readXml`, and getting it from the rw context seems the natural way to do it.

nyalldawson reviewed (commented):
Brilliant. Looking fantastic!

The only thing I think may be missing is that I think we should update layer transform contexts when they are added to a project. E.g. if I create a layer elsewhere, and then later add it to the project, it should have any stored transform context updated/replaced with the project's current one. Do you agree?

Otherwise, I'd like to see unit tests for:
- add layers to a project, check that they auto inherit project context
- change the project context, check that project layer's correctly get the new context

I think we're close to merging here.

nyalldawson committed:
Update src/core/layout/qgslayoutitemmapoverview.cpp

Co-Authored-By: elpaso <elpaso@itopen.it>

nyalldawson committed:
Update src/core/mesh/qgsmeshlayer.h

Co-Authored-By: elpaso <elpaso@itopen.it>

nyalldawson committed:
Update src/core/qgsmaplayer.h

Co-Authored-By: elpaso <elpaso@itopen.it>

elpaso committed:
Pass transform context to raster calculator

elpaso committed:
Use transform context from rw context

elpaso committed:
Drop comment

elpaso committed:
Use new QgsRasterCalculator ctor with transform context

elpaso committed:
doxy

nyalldawson reviewed (commented):
New commits look good

nyalldawson commented:
@elpaso - apart from the things already discussed here, is there anything else left on your todo list?

elpaso commented:
> @elpaso - apart from the things already discussed here, is there anything else left on your todo list?

The tests only.

elpaso commented on src/app/qgisapp.cpp:
```diff
@@ -5768,7 +5768,15 @@ void QgisApp::showRasterCalculator()
   {
     //invoke analysis library
-    QgsRasterCalculator rc( d.formulaString(), d.outputFile(), d.outputFormat(), d.outputRectangle(), d.outputCrs(), d.numberOfColumns(), d.numberOfRows(), QgsRasterCalculatorEntry::rasterEntries() );
+    QgsRasterCalculator rc( d.formulaString(),
```
Isn't it in https://github.com/qgis/QGIS/pull/9627/files?file-filters%5B%5D=.py#diff-3bce79602309fdc721a5c3550a66c958R188 ?

nyalldawson replied:
You're right, sorry

elpaso committed:
Pass transformContext to raster writer

elpaso committed:
Update layer transform context when adding them to a project

elpaso committed:
Set transform context from layer options

elpaso committed:
Get transform context from project instance in app

elpaso committed:
Add transform context to raster file writer task

elpaso committed:
Unit tests for transform context project<->layer sync

elpaso committed:
Fix doxy deprecated

elpaso committed:
Don't use deprecated fileWriter.writeRaster

elpaso committed:
Get transformContext from layer instead of provider

nyalldawson commented:
@elpaso merge away!
